pkg/engine/plugins.go

```diff
@@ -0,0 +1,191 @@
+	}
+	for _, plug := range langhostPlugins {
+		logging.V(preparePluginLog).Infof(
+			"gatherPluginsFromProgram(): plugin %s %s is required by language host", plug.Name, plug.Version)
+		set.Add(plug)
+	}
+	return set, nil
+}
+
+// gatherPluginsFromSnapshot inspects the snapshot associated with the given Target and returns the set of plugins
+// required to operate on the snapshot. The set of plugins is derived from first-class providers saved in the snapshot
+// and the plugins specified in the deployment manifest.
+func gatherPluginsFromSnapshot(plugctx *plugin.Context, target *deploy.Target) (pluginSet, error) {
+	logging.V(preparePluginLog).Infof("gatherPluginsFromSnapshot(): gathering plugins from snapshot")
+	set := newPluginSet()
+	if target == nil || target.Snapshot == nil {
+		logging.V(preparePluginLog).Infof("gatherPluginsFromSnapshot(): no snapshot available, skipping")
+		return set, nil
+	}
+	for _, res := range target.Snapshot.Resources {
+		urn := res.URN
+		if !providers.IsProviderType(urn.Type()) {
+			logging.V(preparePluginVerboseLog).Infof("gatherPluginsFromSnapshot(): skipping %q, not a provider", urn)
+			continue
+		}
+		pkg := providers.GetProviderPackage(urn.Type())
+		version, err := providers.GetProviderVersion(res.Inputs)
+		if err != nil {
+			return set, err
+		}
+		logging.V(preparePluginLog).Infof(
+			"gatherPluginsFromSnapshot(): plugin %s %s is required by first-class provider %q", pkg, version, urn)
+		set.Add(workspace.PluginInfo{
+			Name:    pkg.String(),
+			Kind:    workspace.ResourcePlugin,
+			Version: version,
+		})
+	}
+	for _, plug := range target.Snapshot.Manifest.Plugins {
+		logging.V(preparePluginLog).Infof(
+			"gatherPluginsFromSnapshot(): plugin %s %s is required by snapshot manifest", plug.Name, plug.Version)
+		set.Add(plug)
+	}
+	return set, nil
+}
+
+// ensurePluginsAreInstalled inspects all plugins in the plugin set and, if any plugins are not currently installed,
+// uses the given backend client to install them. Installations are processed in parallel, though
+// ensurePluginsAreInstalled does not return until all installations are completed.
+func ensurePluginsAreInstalled(client deploy.BackendClient, plugins pluginSet) error {
+	if client == nil {
+		logging.V(preparePluginLog).Infoln("ensurePluginsAreInstalled(): skipping due to nil client")
+		return nil
+	}
+	logging.V(preparePluginLog).Infof("ensurePluginsAreInstalled(): beginning")
+	var installTasks errgroup.Group
+	for _, plug := range plugins.Values() {
+		_, path, err := workspace.GetPluginPath(plug.Kind, plug.Name, plug.Version)
+		if err == nil && path != "" {
+			logging.V(preparePluginLog).Infof(
+				"ensurePluginsAreInstalled(): plugin %s %s already installed", plug.Name, plug.Version)
+			continue
+		}
+
+		// Launch an install task asynchronously and add it to the current error group.
+		//
+		// To avoid closing over a loop induction variable by reference, we instruct the Go compiler to make a copy of
```

Comment:
Might be clearer to just manually copy the var into the inner scope:

```
 			info := plug
 			installTasks.Go(func() error {
 				logging.V(preparePluginLog).Infof(
 					"ensurePluginsAreInstalled(): plugin %s %s not installed, doing install", plug.Name, plug.Version)
 				return installPlugin(client, info)
 			})
```

Reply:
I'll give this a try - `info` is still captured by reference but as long as it's not the loop induction variable, it should work?

Reply:
Correct.